pkg/controller/kfservice/kfservice_controller.go

```diff
@@ -123,43 +134,72 @@ func (r *ReconcileService) Reconcile(request reconcile.Request) (reconcile.Resul
+		}
+	}
 
 	serviceReconciler := ksvc.NewServiceReconciler(r.Client)
 
-	ksvc, err := serviceReconciler.Reconcile(context.TODO(), desiredService)
+	defaultConfiguration, err := serviceReconciler.Reconcile(context.TODO(), desiredDefault)
+	if err != nil {
+		log.Error(err, "Failed to reconcile default model spec", "name", desiredDefault.Name)
+		r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	var canaryConfiguration *knservingv1alpha1.Configuration
+	if desiredCanary != nil {
+		canaryConfiguration, err = serviceReconciler.Reconcile(context.TODO(), desiredCanary)
+		if err != nil {
+			log.Error(err, "Failed to reconcile canary model spec", "name", desiredDefault.Name)
+			r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
+			return reconcile.Result{}, err
+		}
+	}
+
+	desiredRoute := resources.CreateKnativeRoute(kfsvc)
+	if err := controllerutil.SetControllerReference(kfsvc, desiredRoute, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+	route, err := serviceReconciler.ReconcileRoute(context.TODO(), desiredRoute)
 	if err != nil {
-		log.Error(err, "Failed to reconcile Knative Serving Service", "name", desiredService.Name)
+		log.Error(err, "Failed to reconcile route", "name", desiredRoute.Name)
 		r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
 		return reconcile.Result{}, err
 	}
 
-	if err = r.updateStatus(kfsvc, ksvc); err != nil {
+	if err = r.updateStatus(kfsvc, defaultConfiguration, canaryConfiguration, route); err != nil {
 		r.Recorder.Eventf(kfsvc, v1.EventTypeWarning, "InternalError", err.Error())
 	}
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileService) updateStatus(before *kfservingv1alpha1.KFService, ksvc *knservingv1alpha1.Service) error {
+func (r *ReconcileService) updateStatus(before *kfservingv1alpha1.KFService,
+	defaultConfiguration *knservingv1alpha1.Configuration,
+	canaryConfiguration *knservingv1alpha1.Configuration,
+	route *knservingv1alpha1.Route) error {
+
 	after := before.DeepCopy()
-	if ksvc.Status.Address != nil {
-		after.Status.URI.Internal = ksvc.Status.Address.Hostname
+	after.Status.Default.Name = defaultConfiguration.Status.LatestCreatedRevisionName
+	if route.Status.Address != nil {
+		after.Status.URI.Internal = route.Status.Address.Hostname
 	}
-	if before.Spec.Canary == nil || before.Spec.Canary.TrafficPercent == 0 {
-		after.Status.Default.Name = ksvc.Status.LatestCreatedRevisionName
-		after.Status.Canary.Name = ""
-	} else {
-		after.Status.Canary.Name = ksvc.Status.LatestCreatedRevisionName
+	if canaryConfiguration != nil {
+		after.Status.Canary.Name = canaryConfiguration.Status.LatestCreatedRevisionName
 	}
-	for _, traffic := range ksvc.Status.Traffic {
+
+	for _, traffic := range route.Status.Traffic {
 		switch traffic.RevisionName {
 		case after.Status.Default.Name:
 			after.Status.Default.Traffic = traffic.Percent
```

Comment:
so we will report traffic percent only for default? so if I set canary.trafficPercent=10, status will show me 90 for default?

Reply:
it is going through the loop from `route.Status.Traffic` which should have both default and canary if canary is set.